addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/ThirdGeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,767 @@
+        System.arraycopy(data, 0, ciphertext, 0, data.length - AES_GCM_TAG_LENGTH / 8);
+        System.arraycopy(data, data.length - AES_GCM_TAG_LENGTH / 8, gcmTag, 0, AES_GCM_TAG_LENGTH / 8);
+
+        JsonObject createSessionJsonObject = new JsonObject();
+        createSessionJsonObject.addProperty("transactionId", transactionId);
+        createSessionJsonObject.addProperty("iv", Base64.getEncoder().encodeToString(iv));
+        createSessionJsonObject.addProperty("tag", Base64.getEncoder().encodeToString(gcmTag));
+        createSessionJsonObject.addProperty("payload", Base64.getEncoder().encodeToString(ciphertext));
+
+        /*
+         * finally create the session for further communication
+         */
+        ContentResponse createSessionResponseContentResponse;
+        JsonObject createSessionResponseJsonObject;
+        try {
+            createSessionResponseContentResponse = executeHttpPost(AUTH_CREATE_SESSION, createSessionJsonObject);
+            createSessionResponseJsonObject = getJsonObjectFromResponse(createSessionResponseContentResponse);
+        } catch (Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    COMMUNICATION_ERROR_AUTHENTICATION);
+            return;
+        }
+        /*
+         * 200 is the desired status code
+         */
+        if (createSessionResponseContentResponse.getStatus() == 400) {
+            /*
+             * Authentication failed
+             */
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    CONFIGURATION_ERROR_PASSWORD);
+            return;
+        }
+
+        sessionId = createSessionResponseJsonObject.get("sessionId").getAsString();
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    /**
+     * Helper function to execute a HTTP post request
+     *
+     * @param resource   web API resource to post to
+     * @param parameters the JSON content to post
+     * @return the HTTP response for the created post request
+     * @throws Exception thrown if there are communication problems
+     */
+    private ContentResponse executeHttpPost(String resource, JsonObject parameters) throws Exception {
+        return executeHttpPost(resource, parameters, null);
+    }
+
+    /**
+     * Helper function to execute a HTTP post request
+     *
+     * @param resource   web API resource to post to
+     * @param sessionId  optional session ID
+     * @param parameters the JSON content to post
+     * @return the HTTP response for the created post request
+     * @throws ClientProtocolException thrown if the web service did not respond correctly
+     * @throws IOException             thrown if there are communication problems
+     */
+    private ContentResponse executeHttpPost(String resource, JsonElement parameters, @Nullable String sessionId)
+            throws Exception {
+        HttpClient client = new HttpClient();
```

Comment:
Don't create a HttpClient, but use `HttpUtil` or inject `HttpClientFactory` via the factory class and get the common client and pass that to your handler (See other bindings to how this is done, just search for `HttpClientFactory` in this repository)

Reply:
I tried using the HttpClientFactory but could not get it to work.
The samples I found in this repository were not helpful :-(
for example: https://github.com/openhab/openhab2-addons/blob/c3bdff9f41c51e943976680c6735bd26b31b6709/addons/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200HandlerFactory.java

There is no clue where this factory is comming from… seems to appearing out of nowhere … 
Is this something the Framework manages if these methods are present?
I don´t get it...

Reply:
> Is this something the Framework manages if these methods are present?
Yes. It's the osgi version of dependency injection. If you put the following lines in a class that has the `@component` annotation (typical the Factory class in a binding) It will be called by the framework and it will set the field. You can then pass it on to the handlers when they are created in the factory:

```
   @Reference
    protected void setHttpClientFactory(HttpClientFactory httpClientFactory) {
        this.httpClient = httpClientFactory.getCommonHttpClient();
    }

    protected void unsetHttpClientFactory(HttpClientFactory httpClientFactory) {
        this.httpClient = null;
    }
```

Reply:
Fixed.